File: subprojects/xsts/xsts-analysis/src/main/java/hu/bme/mit/theta/xsts/analysis/config/XstsTracegenConfig.java
package hu.bme.mit.theta.xsts.analysis.config;

import hu.bme.mit.theta.analysis.Prec;
import hu.bme.mit.theta.analysis.algorithm.SafetyResult;
import hu.bme.mit.theta.analysis.algorithm.tracegeneration.TraceGenerationChecker;
import hu.bme.mit.theta.analysis.expr.ExprState;
import hu.bme.mit.theta.analysis.expr.StmtAction;

import java.util.List;

public class XstsTracegenConfig<S extends ExprState, A extends StmtAction, P extends Prec> {

    private final TraceGenerationChecker<S, A, P> checker;
    private final P initPrec;

    private XstsTracegenConfig(final TraceGenerationChecker<S, A, P> checker, final P initPrec) {
        this.checker = checker;
        this.initPrec = initPrec;
    }

    public static <S extends ExprState, A extends StmtAction, P extends Prec> XstsTracegenConfig<S, A, P> create(
            final TraceGenerationChecker<S, A, P> checker, final P initPrec) {
        return new XstsTracegenConfig<S,A,P>(checker, initPrec);
    }

    public SafetyResult<S, A> check() {
        return checker.check(initPrec);
    }
}

File: subprojects/xsts/xsts-analysis/src/main/java/hu/bme/mit/theta/xsts/analysis/tracegeneration/XstsTracegenConfig.kt
/*
 *  Copyright 2024-2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.xsts.analysis.tracegeneration

import hu.bme.mit.theta.analysis.Action
import hu.bme.mit.theta.analysis.Prec
import hu.bme.mit.theta.analysis.State
import hu.bme.mit.theta.analysis.algorithm.tracegeneration.CegarTraceGenerationChecker
import hu.bme.mit.theta.analysis.algorithm.tracegeneration.summary.AbstractTraceSummary
import hu.bme.mit.theta.analysis.algorithm.tracegeneration.summary.TraceGenerationResult

class XstsTracegenConfig<S : State, A : Action, P : Prec>
private constructor(
  private val checker: CegarTraceGenerationChecker<S, A, P>,
  private val prec: P,
) {
  fun check(): TraceGenerationResult<AbstractTraceSummary<S, A>, S, A> {
    return checker.check(prec)
  }

  companion object {
    fun <S : State, A : Action, P : Prec> create(
      checker: CegarTraceGenerationChecker<S, A, P>,
      initPrec: P,
    ): XstsTracegenConfig<S, A, P> {
      return XstsTracegenConfig(checker, initPrec)
    }
  }
}
